apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,366 @@
+        require(_vestingCompletePeriod > _vestingCliffPeriod, ERROR_INVALID_TIME_PERIOD);
+        require(_fundingGoal > 0, ERROR_INVALID_FUNDING_GOAL);
+        require(_percentSupplyOffered > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentSupplyOffered < PPM, ERROR_INVALID_PERCENT_VALUE);
+        require(_beneficiaryAddress != 0x0, ERROR_INVALID_BENEFIC_ADDRESS);
+        require(_percentFundingForBenefiriary > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentFundingForBenefiriary < PPM, ERROR_INVALID_PERCENT_VALUE);
+
+        initialized();
+
+        contributionToken = _contributionToken;
+        _setProjectToken(_projectToken, _projectTokenManager);
+
+        fundraisingPool = _fundraisingPool;
+
+        vestingCliffPeriod = _vestingCliffPeriod;
+        vestingCompletePeriod = _vestingCompletePeriod;
+        fundingPeriod = _fundingPeriod;
+
+        beneficiaryAddress = _beneficiaryAddress;
+        percentFundingForBeneficiary = _percentFundingForBenefiriary;
+
+        fundingGoal = _fundingGoal;
+        percentSupplyOffered = _percentSupplyOffered;
+
+        if (_startDate != 0) {
+            _setStartDate(_startDate);
+        }
+
+        _calculateExchangeRate();
+    }
+
+    /*
+     * Public
+     */
+
+    /**
+    * @notice Starts the sale, changing its state to Funding. After the sale is started contributors will be able to purchase project tokens.
+    */
+    function start() public auth(START_ROLE) {
+        require(currentSaleState() == SaleState.Pending, ERROR_INVALID_STATE);
+        _setStartDate(getTimestamp64());
+    }
+
+    /**
+    * @notice Buys project tokens using the provided `_tokensToSpend` contribution tokens. To calculate how many project tokens will be sold for the provided contribution tokens amount, use contributionToProjectTokens(). Each purchase generates a numeric vestedPurchaseId (0, 1, 2, etc) for the caller, which can be obtained in the TokensPurchased event emitted, and is required for later refunds. Note: If `_tokensToSpend` + `totalRaised` is larger than `fundingGoal`, only part of it will be used so that the funding goal is never exceeded.
+    * @param _tokensToSpend The amount of contribution tokens to spend to obtain project tokens.
+    */
+    function buy(uint256 _tokensToSpend) public auth(BUY_ROLE) {
+        require(currentSaleState() == SaleState.Funding, ERROR_INVALID_STATE);
+
+        uint256 tokensToUse = _tokensToSpend;
+        if (totalRaised.add(tokensToUse) > fundingGoal) {
+            tokensToUse = fundingGoal.sub(totalRaised);
+        }
+
+        require(contributionToken.balanceOf(msg.sender) >= tokensToUse, ERROR_INSUFFICIENT_BALANCE);
+        require(contributionToken.allowance(msg.sender, address(this)) >= tokensToUse, ERROR_INSUFFICIENT_ALLOWANCE);
+
+        // (buyer) ~~~> contribution tokens ~~~> (presale)
+        require(contributionToken.transferFrom(msg.sender, address(this), tokensToUse), ERROR_TOKEN_TRANSFER_REVERTED);
```

Comment:
@osarrouy 

Well, can we guarantee that the contribution token will be SafeERC20 compliant? It would have to implement the interface for `safeTransfer`, `safeTransferFrom`, `staticBalanceOf` etc.

Reply:
`SafeERC20` is a lib so I think the point is that we can use it on top of any ERC20 [to make sure it actually behaves like an ERC20]. You can see the `BatchedBancorMarketMaker `contract if you wanna have a look ?

Reply:
Ah! Of course :)